claat/fetch/drive/auth/auth.go

```diff
@@ -37,117 +35,107 @@ const (
 	googSecret = "ljELuf5jUrzcOxZGL7OQfkIC"
 
 	// token providers
-	providerGoogle = "goog"
+	ProviderGoogle = "goog"
 )
 
 var (
-	// OAuth2 configs for OOB flow
-	authConfig = map[string]*oauth2.Config{
-		providerGoogle: {
-			ClientID:     googClient,
-			ClientSecret: googSecret,
-			Scopes:       []string{scopeDriveReadOnly},
-			RedirectURL:  "urn:ietf:wg:oauth:2.0:oob",
-			Endpoint: oauth2.Endpoint{
-				AuthURL:  "https://accounts.google.com/o/oauth2/auth",
-				TokenURL: "https://accounts.google.com/o/oauth2/token",
-			},
+	googleAuthConfig = oauth2.Config{
+		ClientID:     googClient,
+		ClientSecret: googSecret,
+		Scopes:       []string{scopeDriveReadOnly},
+		RedirectURL:  "urn:ietf:wg:oauth:2.0:oob",
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  "https://accounts.google.com/o/oauth2/auth",
+			TokenURL: "https://accounts.google.com/o/oauth2/token",
 		},
 	}
-
-	// reusable HTTP clients
-	clientsMu sync.Mutex // guards clients
-	clients   map[string]*http.Client
 )
 
-func init() {
-	clients = make(map[string]*http.Client)
+type authorizationHandler func(conf *oauth2.Config) (*oauth2.Token, error)
+
+type internalOptions struct {
+	authHandler authorizationHandler
 }
 
-// driveClient returns an HTTP client which knows how to perform authenticated
-// requests to Google Drive API.
-func driveClient(authToken string) (*http.Client, error) {
-	clientsMu.Lock()
-	defer clientsMu.Unlock()
-	provider := providerGoogle + authToken;
-	if hc, ok := clients[provider]; ok {
-		return hc, nil
+type Helper struct {
+	authToken string
+	provider  string
+	client    *http.Client
+	opts      internalOptions
+}
+
+func NewHelper(at, p string, rt http.RoundTripper) (*Helper, error) {
+	return newHelper(at, p, rt, internalOptions{
+		authHandler: authorize,
+	})
+}
+
+func newHelper(at, p string, rt http.RoundTripper, io internalOptions) (*Helper, error) {
+	h := Helper{
+		authToken: at,
+		provider:  p,
+		opts:      io,
 	}
-	ts, err := tokenSource(providerGoogle, authToken)
+
+	var err error
+	h.client, err = h.produceDriveClient(rt)
 	if err != nil {
 		return nil, err
 	}
-	t := &oauth2.Transport{
-		Source: ts,
-		Base:   http.DefaultTransport,
+
+	return &h, nil
+}
+
+func (h *Helper) DriveClient() *http.Client {
+	return h.client
+}
+
+func (h *Helper) produceDriveClient(rt http.RoundTripper) (*http.Client, error) {
+	ts, err := h.tokenSource()
+	if err != nil {
+		return nil, err
+	}
+
+	if rt == nil {
+		rt = http.DefaultTransport
 	}
-	hc := &http.Client{Transport: t}
-	clients[provider] = hc
-	return hc, nil
+
+	return &http.Client{
+		Transport: &oauth2.Transport{
+			Source: ts,
+			Base:   rt,
+		},
+	}, nil
 }
 
 // tokenSource creates a new oauth2.TokenSource backed by tokenRefresher,
 // using previously stored user credentials if available.
-// If authToken is given, we disregard the value of provider.
+// If authToken is not given, we use the Google provider.
```

Comment:
"not given" -> "not given in the constructor/new", since this itself doesn't take in parameters.